refactor: routes as virtual module

ematipico opened the pull request:
## Changes

This PR adds a new virtual module that allows to return the list of routes inside an Astro project. 

This virtual module essentially replaces the `RoutesList` type that we create when we start the dev server.

## Testing

Unit tests should pass. Integration tests pass on my machine, they will still time out here

<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->

## Docs

N/A

<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
<!-- /cc @withastro/maintainers-docs for feedback! -->

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

ascorbic reviewed (commented):
Will `App` be able to load this if there are no routes in the manifest?

florian-lefebvre commented on packages/astro/src/vite-plugin-routes/index.ts:
```diff
@@ -0,0 +1,78 @@
+		name: 'astro:routes',
+		enforce: 'pre',
+		configureServer(server) {
+			server.watcher.on('all', (_event, relativeEntry) => {
+				const entry = new URL(relativeEntry, settings.config.root);
+				if (entry.pathname.startsWith(settings.config.srcDir.pathname)) {
+					server.restart();
+				}
+			});
```
It happens here I believe: https://github.com/withastro/astro/blob/main/packages/astro/src/vite-plugin-astro-server/plugin.ts#L71-L105

florian-lefebvre replied:
So maybe instead of restarting, you could re-call the `astro:routes` plugin?

ematipico replied:
Actually, it wasn't working at all. I refactored it, and we now use the same logic as on the dev server. It works as expected now, I tested it locally

ematipico commented on packages/astro/src/vite-plugin-astro-server/app.ts:
```diff
@@ -57,7 +57,6 @@ export class DevApp extends BaseApp<DevPipeline> {
 		this.manifestData = manifestData;
 	}
 
-	// TODO remove routelist once it's a plugin
```
Yes

ematipico commented:
> [!IMPORTANT]
> I removed the `astro:scanner` plugin, and moved its logic inside the new `astro:routes` plugin. Both plugins need to work around the routes that are created from `createRoutesList`, and leaving them as two separate plugins was a problem. Merging them made things easier.

ematipico commented:
The problem is caused now by an inconsistency between the dev/ssr pipeline and the build pipeline. 

Particularly, there's a plugin called `plugin-manifest`, which is a plugin excessively used during the build.

I will address this in a separate PR.
